From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Euphyllia Bierque <bierque.euphyllia@gmail.com>
Date: Tue, 28 Jan 2025 16:35:13 -0800
Subject: [PATCH] Add TPS From Region


diff --git a/src/main/java/io/papermc/paper/command/PaperCommands.java b/src/main/java/io/papermc/paper/command/PaperCommands.java
index 3deca0b4e95609dc95376a86b5cbd525729a1031..cf9996228b320dd466de706928aa75df6df4340f 100644
--- a/src/main/java/io/papermc/paper/command/PaperCommands.java
+++ b/src/main/java/io/papermc/paper/command/PaperCommands.java
@@ -25,7 +25,6 @@ public final class PaperCommands {
         COMMANDS.put("paper", new PaperCommand("paper"));
         COMMANDS.put("mspt", new MSPTCommand("mspt"));
         COMMANDS.put("tps", new io.papermc.paper.threadedregions.commands.CommandServerHealth()); // Folia - region threading
-        COMMANDS.put("profiler", new io.papermc.paper.threadedregions.commands.CommandProfiler()); // Folia - region threading - profiler
 
         COMMANDS.forEach((s, command) -> {
             server.server.getCommandMap().register(s, "Paper", command);
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index eebe5b9126e1771df41ee6efb9e3d3ab313c18c0..a2d9343b99093ade5e79cf6e5cf98e4fce8f2f19 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -3032,4 +3032,69 @@ public final class CraftServer implements Server {
     public void allowPausing(final Plugin plugin, final boolean value) {
         this.console.addPluginAllowingSleep(plugin.getName(), value);
     }
+
+    // Folia start - region TPS API
+    /**
+     * Gets the TPS from the region which owns the specified location, or {@code null} if no region owns
+     * the specified location.
+     *
+     * @param location The location for which to get the TPS
+     * @return TPS (5s, 15s, 1m, 5m, 15m), or null if the region doesn't exist
+     */
+    @Override
+    public double[] getRegionTPS(Location location) {
+        Preconditions.checkArgument(location != null, "Location cannot be null");
+
+        return this.getRegionTPS(location.getWorld(), location.getBlockX() >> 4, location.getBlockZ() >> 4);
+    }
+
+    /**
+     * Gets the TPS from the region which owns the specified chunk, or {@code null} if no region owns
+     * the specified location.
+     *
+     * @param chunk - The specified chunk
+     * @return TPS (5s, 15s, 1m, 5m, 15m), or null if the region doesn't exist
+     */
+    @Override
+    public double[] getRegionTPS(org.bukkit.Chunk chunk) {
+        Preconditions.checkArgument(chunk != null, "Chunk cannot be null");
+
+        return this.getRegionTPS(chunk.getWorld(), chunk.getX(), chunk.getZ());
+    }
+
+    /**
+     * Gets the TPS from the region which owns the specified chunk, or {@code null} if no region owns
+     * the specified location.
+     *
+     * @param world - World containing the chunk
+     * @param chunkX - X-coordinate of the chunk
+     * @param chunkZ - Z-coordinate of the chunk
+     * @return TPS (5s, 15s, 1m, 5m, 15m), or null if the region doesn't exist
+     */
+    @Override
+    public double[] getRegionTPS(World world, int chunkX, int chunkZ) {
+        Preconditions.checkArgument(world != null, "World cannot be null");
+
+        return getTPSFromRegion(((CraftWorld)world).getHandle(), chunkX, chunkZ);
+    }
+
+    private static double[] getTPSFromRegion(ServerLevel world, int chunkX, int chunkZ) {
+        io.papermc.paper.threadedregions.ThreadedRegionizer.ThreadedRegion<io.papermc.paper.threadedregions.TickRegions.TickRegionData, io.papermc.paper.threadedregions.TickRegions.TickRegionSectionData>
+                region = world.regioniser.getRegionAtSynchronised(chunkX, chunkZ);
+        if (region == null) {
+            return null;
+        } else {
+            io.papermc.paper.threadedregions.TickRegions.TickRegionData regionData = region.getData();
+            final long currTime = System.nanoTime();
+            final io.papermc.paper.threadedregions.TickRegionScheduler.RegionScheduleHandle regionScheduleHandle = regionData.getRegionSchedulingHandle();
+            return new double[] {
+                    regionScheduleHandle.getTickReport5s(currTime).tpsData().segmentAll().average(),
+                    regionScheduleHandle.getTickReport15s(currTime).tpsData().segmentAll().average(),
+                    regionScheduleHandle.getTickReport1m(currTime).tpsData().segmentAll().average(),
+                    regionScheduleHandle.getTickReport5m(currTime).tpsData().segmentAll().average(),
+                    regionScheduleHandle.getTickReport15m(currTime).tpsData().segmentAll().average(),
+            };
+        }
+    }
+    // Folia end - region TPS API
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlockState.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlockState.java
index b19c1f90adf5ef4824fb6f106b9afa733d4ee7ae..8804538cacbfcae1430faf8d01b40c2c4713726b 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlockState.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlockState.java
@@ -215,7 +215,7 @@ public class CraftBlockState implements BlockState {
         }
 
         // Folia start - region threading
-        if (access instanceof net.minecraft.server.level.ServerLevel serverWorld) {
+        if (this.getWorldHandle() instanceof net.minecraft.server.level.ServerLevel serverWorld) {
             ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(serverWorld, position, "Cannot modify world asynchronously");
         }
         // Folia end - region threading
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
index de0e88caf608562cf004877ac693bfe4130598a6..1fb035f632b07913e2a611aa45e0972e8affb782 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
@@ -161,7 +161,7 @@ public final class CraftItemStack extends ItemStack {
     /**
      * Copies the NMS stack to return as a strictly-Bukkit stack
      */
-    private static ItemStack asBukkitCopy(net.minecraft.world.item.ItemStack original) {
+    public static ItemStack asBukkitCopy(net.minecraft.world.item.ItemStack original) { // Folia - plugin compat: keep this reflectable (cloud/Geyser look it up via getMethod)
         // no such thing as a "strictly-Bukkit stack" anymore
         // we copy the stack since it should be a complete copy not a mirror
         return asCraftMirror(original.copy());
